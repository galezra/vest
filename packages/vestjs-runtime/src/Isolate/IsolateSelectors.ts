import { Maybe } from 'vest-utils';

import { TIsolate } from 'Isolate';
import { IsolateKeys } from 'IsolateKeys';

export function isIsolateType<I extends TIsolate>(
  node: Maybe<TIsolate>,
  type: string,
): node is I {
  return node?.[IsolateKeys.Type] === type;
}

export function isSameIsolateType<A extends TIsolate, B extends TIsolate>(
  a: A,
  b: B,
): boolean {
  return isIsolateType(a, b[IsolateKeys.Type]);
}

export function isSameIsolateIdentity<A extends TIsolate, B extends TIsolate>(
  a: A,
  b: B,
): boolean {
  return Object.is(a, b) || (isSameIsolateType(a, b) && a.key === b.key);
}
