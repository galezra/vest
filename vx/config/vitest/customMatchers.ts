import { isFunction } from 'vest-utils';

expect.extend({
  isDeepCopyOf,
});

// eslint-disable-next-line complexity
function isDeepCopyOf(
  source: any,
  clone: any,
): { pass: boolean; message: () => string } {
  const queue = [[source, clone]];

  outer: while (queue.length) {
    const [source, clone] = (queue as any[]).shift();

    if (!source || typeof source !== 'object') {
      if (!isFunction(clone)) {
        if (clone !== source) {
          return {
            pass: false,
            message: () => `Source and clone are not identical`,
          };
        }
      }
      continue;
    }

    if (clone === source) {
      return {
        pass: false,
        message: () =>
          `Source and clone are the same object. Expected a deep copy. ${JSON.stringify(
            source,
          )}===${JSON.stringify(clone)}`,
      };
    }

    // Short circuit
    if (
      (clone && !source) ||
      (source && !clone) ||
      typeof source !== typeof clone
    ) {
      return {
        pass: false,
        message: () =>
          `Source and clone are not of the same type: ${JSON.stringify(
            source,
          )} does not equal ${JSON.stringify(clone)}`,
      };
    }

    if (Array.isArray(source)) {
      // Short circuit
      if (!Array.isArray(clone) || source.length !== clone.length) {
        return {
          pass: false,
          message: () =>
            `source and clone arrays are not identical. ${JSON.stringify(
              source,
            )} does not equal ${JSON.stringify(clone)}`,
        };
      }
      source.forEach((_, i) => {
        queue.push([source[i], clone[i]]);
      });
    } else if (typeof source === 'object') {
      Object.keys(source).forEach(key => queue.push([source[key], clone[key]]));
    }

    continue outer;
  }

  return { pass: true, message: () => 'success' };
}
