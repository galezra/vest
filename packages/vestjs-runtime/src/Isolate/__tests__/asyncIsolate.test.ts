import { useBus } from 'Bus';
import { Isolate, TIsolate } from 'Isolate';
import { RuntimeEvents } from 'RuntimeEvents';
import { CB } from 'vest-utils';
import { describe, vi, it, expect, test } from 'vitest';
import wait from 'wait';

import { VestRuntime } from 'vestjs-runtime';

describe('AsyncIsolate', () => {
  test('It should resolve async isolate into the parent', () => {
    return new Promise<void>(async done => {
      let root = {} as TIsolate;
      withRunTime(() => {
        // Create root isolate from which all others will be created
        root = Isolate.create('URoot', genChildren);
      });
      expect(root).toMatchInlineSnapshot(`
        {
          "$type": "URoot",
          "abortController": AbortController {},
          "allowReorder": undefined,
          "children": null,
          "data": {},
          "key": null,
          "keys": null,
          "output": Promise {},
          "parent": null,
        }
      `);
      await wait(10);
      expect(root?.children?.[0]?.$type).toBe('UChild_1');
      expect(root?.children?.[0].parent).toBe(root);
      expect(root?.children?.[0]?.children?.[0]?.$type).toBe('UGrandChild_1');
      expect(root?.children?.[0]?.children?.[0].parent).toBe(
        root?.children?.[0],
      );
      expect(root?.children?.[0]?.children?.[1]?.$type).toBe('UGrandChild_2');
      expect(root?.children?.[0]?.children?.[1].parent).toBe(
        root?.children?.[0],
      );
      expect(root?.children?.[0]?.children?.[2]?.$type).toBe('UGrandChild_3');
      expect(root?.children?.[0]?.children?.[2].parent).toBe(
        root?.children?.[0],
      );
      expect(root).toMatchSnapshot();

      done();
    });
  });

  it('Should emit an event when an async isolate is done running', () => {
    const cb = vi.fn();
    return new Promise<void>(async done => {
      let child = {} as TIsolate;
      withRunTime(() => {
        // Create root isolate from which all others will be created
        Isolate.create('URoot', () => {
          const bus = useBus();
          bus.on(RuntimeEvents.ISOLATE_DONE, cb);

          expect(cb).not.toHaveBeenCalled();
          child = Isolate.create('UChild_1', async () => {
            await wait(10);
          });
          expect(cb).not.toHaveBeenCalled();
        });
      });
      expect(cb).not.toHaveBeenCalledWith(child);
      await wait(10);
      expect(cb).toHaveBeenCalledWith(child);
      done();
    });
  });
});

function withRunTime<T>(fn: CB<T>) {
  return VestRuntime.Run(VestRuntime.createRef({}), () => {
    return fn();
  });
}

async function genChildren() {
  await wait(10);
  // Create first child isolate
  return withRunTime(() =>
    Isolate.create('UChild_1', () => {
      // Create three grandchildren
      Isolate.create('UGrandChild_1', () => {});
      Isolate.create('UGrandChild_2', () => {});
      Isolate.create('UGrandChild_3', () => {});
    }),
  );
}
