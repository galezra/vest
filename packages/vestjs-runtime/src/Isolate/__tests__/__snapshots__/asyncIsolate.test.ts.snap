// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AsyncIsolate > It should resolve async isolate into the parent 2`] = `
{
  "$type": "URoot",
  "abortController": AbortController {},
  "allowReorder": undefined,
  "children": [
    {
      "$type": "UChild_1",
      "abortController": AbortController {},
      "allowReorder": undefined,
      "children": [
        {
          "$type": "UGrandChild_1",
          "abortController": AbortController {},
          "allowReorder": undefined,
          "children": null,
          "data": {},
          "key": null,
          "keys": null,
          "output": undefined,
          "parent": [Circular],
        },
        {
          "$type": "UGrandChild_2",
          "abortController": AbortController {},
          "allowReorder": undefined,
          "children": null,
          "data": {},
          "key": null,
          "keys": null,
          "output": undefined,
          "parent": [Circular],
        },
        {
          "$type": "UGrandChild_3",
          "abortController": AbortController {},
          "allowReorder": undefined,
          "children": null,
          "data": {},
          "key": null,
          "keys": null,
          "output": undefined,
          "parent": [Circular],
        },
      ],
      "data": {},
      "key": null,
      "keys": null,
      "output": undefined,
      "parent": [Circular],
    },
  ],
  "data": {},
  "key": null,
  "keys": null,
  "output": Promise {},
  "parent": null,
}
`;
